import os
import argparse
from pathlib import Path
import pandas as pd

ROOT_DIR = Path.cwd()
EXPERIMENTS_DIR=f"{ROOT_DIR}/experiments"
TREEMIX=f"{os.environ['HOME']}/miniconda3/envs/treemix_env/bin/treemix"
SEED=1
SNP_SIZE=1000
SPLIT_ONLY=1
MIGRATION_ENABLED=10
ROOT="San"


EXPERIMENTS = {
    # Experiment 1: Baseline (Split only and migration enabled)
    "baseline": {
        "plink_test": "test_1",
        "plink_output":"baseline/plink_results/baseline_treemix.gz",
        "variants": []
    },

    # Experiment 2a: Treemix Parameters - SNP Block Size
    "exp_2a": {
        "plink_test":"test_1", # Use the baseline PLINK file
        "plink_output":"baseline/plink_results/baseline_treemix.gz",
        "variants":[
            {"name":"k200","k":200},
            {"name":"k500","k":500},
            {"name":"k1000","k":1000},
            {"name":"k1500","k":1500},
            {"name":"k2000","k":2000},
            {"name":"k10000","k":10000}
        ]
    },

    # Experiment 2b: Treemix Parameters - Migration Edges
    "exp_2b": {
        "plink_test": "test_1", #Use the baseline PLINK file
        "plink_output": "baseline/plink_results/baseline_treemix.gz",
        "variants":[
            {"name":f"m{m}","m":m}
            for m in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
        ]
    },

    # Experiment 3a: PLINK Parameters - Genotype missigness
    "exp_3a": {
        "plink_test":"test_3a",
        "variants":[
            {"name":"geno_missing_000","geno":"0.00"},
            {"name":"geno_missing_001","geno":"0.01"},
            {"name":"geno_missing_005","geno":"0.05"}
        ]
    },

    # Experiment 3b: PLINK Parameters - Minor Allele Frequency
    "exp_3b": {
        "plink_test":"test_3b",
        "variants": [
            {"name":"minor_allele_freq_005","maf":"0.05"}
        ]
    },

    # Experiment 4: Addtion of Archaic Homonin Genomes
    "exp_4" : {
        "plink_test":"test_4",
        "variants":[
            {"name":"added_Deni"},
            {"name":"added_Vindija"},
            {"name":"added_Both"}
        ]
    },

    # Experiment 5a: Uneven Sampling
    "exp_5a": {
        "plink_test":"test_5a",
        "variants": [ 
            {"name":"filtered_3_uneven_ind","size":3},
            {"name":"filtered_5_uneven_ind","size":5}
        ]
    },

        # Experiment 5b: Even sampling
    "exp_5b": {
        "plink_test": "test_5b",
        "variants": [
            {"name": "filtered_3_even_ind", "size": 3},
            {"name": "filtered_5_even_ind", "size": 5}
        ]
    },
    
    # Experiment 6a: Drop populations
    "exp_6a": {
        "plink_test": "test_6a",
        "variants": [
            {"name": "drop_pops_14", "num": 14},
            {"name": "drop_pops_20", "num": 20}
        ]
    },
    
    # Experiment 6b: Remove continents
    "exp_6b": {
        "plink_test": "test_6b",
        "variants": [
            {"name": f"dataset_no_{continent.replace(' ', '_')}"} 
            for continent in ["Europe", "North_Africa", "Middle_East", "Asia", "America", "Oceania"]
        ]
    },
    
    # Experiment 7a: Remove admixed populations
    "exp_7a": {
        "plink_test": "test_7a",
        "variants": [{"name": "drop_admixed_pops"}]
    },
    
    # Experiment 7ba: Artificial 2-way hybrids
    "exp_7ba": {
        "plink_test": "test_7ba",
        "variants": [{"name": "artificial_2_merged_all"}]
    },
    
    # Experiment 7bb: Artificial 5-way hybrids
    "exp_7bb": {
        "plink_test": "test_7bb",
        "variants": [{"name": "artificial_5_merged_all"}]
    },
    
    # Experiment 8: Artificial shuffled
    #"exp_8": {
    #    "plink_test": "test_8",
    #    "variants": [{"name": "artificial_5_merged_all_shuffled"}]
    #}
}


# ======== Helper Function ========
def get_all_plink_outputs():
    # Get all the plink output file names and store it in a list
    outputs = []
    for exp_name, exp_config in EXPERIMENTS.items():
        if "plink_output" in exp_config.keys():
            write_out = f"{EXPERIMENTS_DIR}/{exp_config['plink_output']}"
            outputs.append(write_out)
        else:
            variant_id = [var_name['name'] for var_name in exp_config['variants']]
            for id in variant_id:
                write_out = f"{EXPERIMENTS_DIR}/experiment_{exp_name.split("_")[-1]}/plink_results/{exp_name.split("_")[-1]}_{id}_treemix.gz"
                outputs.append(write_out)
    return outputs

def get_outputs_for_test(exp_name):
    # Get all the ouput for a particular experiment
    outputs = []
    output = EXPERIMENTS[exp_name]
    for key, value in output.items():
        if key == 'plink_output':
            write_out = f"{EXPERIMENTS_DIR}/{output['plink_output']}"
            outputs.append(write_out)
            break
        elif key == 'variants':
            variant_id = [i['name'] for i in output[key]]
            for id in variant_id:
                write_out = f"{EXPERIMENTS_DIR}/experiment_{exp_name.split("_")[-1]}/plink_results/{exp_name.split("_")[-1]}_{id}_treemix.gz"
                outputs.append(write_out)
    return outputs       

def get_all_treemix_outputs():
    outputs = []
    for exp_name, exp_config in EXPERIMENTS.items():
        if exp_name == "baseline":
            exp_dir = "baseline"
            variant_names = ["baseline"]
            prefix = "baseline"
        elif exp_name in ["exp_2a","exp_2b"]:
            continue
        else:
            exp_dir = f"experiment_{exp_name.split('_')[-1]}"
            prefix = exp_name.split('_')[-1]
            variant_names = [v["name"] for v in exp_config["variants"]]

        for variant in variant_names:
            full_variant = f"{prefix}_{variant}" if prefix != "baseline" else variant
            for mode in ["split", "migration"]:
                for ext in ["treeout.gz", "vertices.gz", "edges.gz"]:
                    outputs.append(
                        f"{EXPERIMENTS_DIR}/{exp_dir}/treemix_results/{full_variant}_{mode}.{ext}"
                    )
    return outputs

# ============ Rules ==================

# Rule all: Define the output files I need
rule all:
    input:
        # PLINK outputs
        #get_all_plink_outputs()
        get_outputs_for_test("baseline"),
        get_outputs_for_test("exp_3a"),
        get_outputs_for_test("exp_3b"),
        get_outputs_for_test("exp_4"),
        get_outputs_for_test("exp_5a"),
        get_outputs_for_test("exp_5b"),
        get_outputs_for_test("exp_6a"),
        get_outputs_for_test("exp_6b"),
        get_outputs_for_test("exp_7a"),
        get_outputs_for_test("exp_7ba"),
        get_outputs_for_test("exp_7bb"),
        #get_outputs_for_test("exp_8"),

        # Treemix outputs - What I need? treeout.gz, vertices.gz and edges.gz for each condition
        #expand(f"{EXPERIMENTS_DIR}/{{exp}}/treemix_results/{{filename}}_{{mode}}.{{ext}}",
        #        mode = ["split","migration"],
        #        ext = ["treeout.gz","vertices.gz","edges.gz"],
        #        exp = [exp.replace("exp_","experiment_") if exp.startswith("exp") else exp for exp in EXPERIMENTS.keys()],
        #        filename = [filename.split('plink_results/')[-1].split('_treemix.gz')[0] for filename in get_all_plink_outputs()])
        get_all_treemix_outputs()

# Step 1: Run the PLINK SCRIPT for each experiment 
rule run_plink_test_1:
    output:
        get_outputs_for_test("baseline")

    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    
    threads: 1
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_1
        """

rule run_plink_test_3a:
        output:
            get_outputs_for_test("exp_3a")
        params:
            plink_script = f"{ROOT_DIR}/src/plink_script.sh"
        threads: 6
        shell:
            """
            export SNAKEMAKE_CORES={threads}
            bash {params.plink_script} test_3a
            """

rule run_plink_test_3b:
        output:
            get_outputs_for_test("exp_3b")
        params:
            plink_script = f"{ROOT_DIR}/src/plink_script.sh"
        threads: 1
        shell:
            """
            export SNAKEMAKE_CORES={threads}
            bash {params.plink_script} test_3b
            """

rule run_plink_test_4:
    output:
        get_outputs_for_test("exp_4")
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    threads: 5
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_4
        """

rule run_plink_test_5a:
    output:
        get_outputs_for_test("exp_5a")
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    threads: 3
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_5a
        """

rule run_plink_test_5b:
    output:
        get_outputs_for_test("exp_5b")
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    threads: 3
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_5b
        """
rule run_plink_test_6a:
    output:
        get_outputs_for_test("exp_6a")
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    threads: 5
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_6a
        """

rule run_plink_test_6b:
    output:
        get_outputs_for_test("exp_6b")
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    threads: 10
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_6b
        """

rule run_plink_test_7a:
    output:
        get_outputs_for_test("exp_7a")

    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    
    threads: 3
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_7a
        """
rule run_plink_test_7ba:
    output:
        get_outputs_for_test("exp_7ba")
    
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    
    threads: 6
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_7ba
        """

rule run_plink_test_7bb:
    output:
        get_outputs_for_test("exp_7bb")
    
    params:
        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
    
    threads: 6
    shell:
        """
        export SNAKEMAKE_CORES={threads}
        bash {params.plink_script} test_7bb
        """

#rule run_plink_test_8:
#    output:
#        get_outputs_for_test("exp_8")
#    
#    params:
#        plink_script = f"{ROOT_DIR}/src/plink_script.sh"
#    
#    threads: 6
#    shell:
#        """
#        export SNAKEMAKE_CORES={threads}
#        bash {params.plink_script} test_8
#        """

# Step 2: Run treemix 

rule run_treemix:
    input:
        treemix_file=f"{EXPERIMENTS_DIR}/{{exp}}/plink_results/{{filename}}_treemix.gz"
    
    output:
        treeout=f"{EXPERIMENTS_DIR}/{{exp}}/treemix_results/{{filename}}_{{mode}}.treeout.gz",
        vertices=f"{EXPERIMENTS_DIR}/{{exp}}/treemix_results/{{filename}}_{{mode}}.vertices.gz",
        edges=f"{EXPERIMENTS_DIR}/{{exp}}/treemix_results/{{filename}}_{{mode}}.edges.gz"
        log=f"{EXPERIMENTS_DIR}/{{exp}}/treemix_results/{{filename}}_{{mode}}.log"
    
    threads: 3
       
    params:
        m = lambda w: 0 if w.mode == 'split' else 10,
        k = SNP_SIZE,
        seed = SEED,
        root = ROOT,
        output_stem = lambda w: f"{EXPERIMENTS_DIR}/{w.exp}/treemix_results/{w.filename}_{w.mode}",
        results_dir = lambda w: f"{EXPERIMENTS_DIR}/{w.exp}/treemix_results"
    
    shell:
        """
        # Create the results directory
        mkdir -p {params.results_dir}


        {TREEMIX} -i {input.treemix_file} \
                    -m {params.m} \
                    -k {params.k} \
                    -seed {params.seed} \
                    -root {params.root} \
                    -o {params.output_stem} > {output.log} 2>&1
        """

# Step 3: Find Differences


# Step 4: Plot the tree